File: 1003_サイコロの実装(1)/main.scala

trait IOProvider {
  def read(): String
  def write(str: String): Unit
}


class StdIOProvider extends IOProvider {
  def read(): String = {
    return io.StdIn.readLine()
  }

  def write(str: String): Unit = {
    println(str)
  }
}

class TestIOProvider extends IOProvider {
  def read(): String = {
    return "6"
  }

  def write(str: String): Unit = {
    println(str)
  }
}


abstract class Base(provider: IOProvider) {
  val io: IOProvider = provider;

  def read(): String = {
    return io.read()
  }

  def write(str: String): Unit = {
    io.write(str)
  }

  def main(): Unit = {}
}


class Program(provider: IOProvider) extends Base(provider) {
  override def main(): Unit = {
    val input: String = this.read()

    val n: Int = input.toInt
    val l: List[Int] = (1 to n).toList.map(_ % 6 + 1)
    val l2: List[Int] = (1 to 6).toList.map(i => l.filter(j => i == j).size)
    val result: Boolean = l2.forall(_ == l2.head)

    this.write(if (result) "Yes" else "No")
  }
}


object Main {
  def main(args: Array[String]): Unit = {
    val provider = new TestIOProvider()
    new Program(provider).main()
  }
}

File: 1003_サイコロの実装(1)/main.dart
import 'dart:io';

List<int> range(int stop, [int start = 0]) {
  return new List<int>.generate(stop, (i) => start + i);
}

abstract class IOProvider {
  String read();
  void write(String s);
}

class StdIOProvider implements IOProvider {
  String read() {
    return stdin.readLineSync();
  }

  void write(String s) {
    print(s);
  }
}

class TestIOProvider implements IOProvider {
  String read() {
    return "6";
  }

  void write(String s) {
    print(s);
  }
}

class Base {
  IOProvider io;

  Base(IOProvider provider) {
    this.io = provider;
  }

  String read() {
    return this.io.read();
  }

  void write(String s) {
    this.io.write(s);
  }
}

class Program extends Base {
  Program(IOProvider provider) : super(provider);

  void main() {
    final String input = read();

    final int n = int.parse(input);
    List<int> l = range(n, 1).map((int i) => i % 6 + 1).toList();
    l = range(6, 1).map((int i) => l.where((int j) => i == j).length).toList();
    bool result = l.every((int i) => i == l[0]);

    write(result ? "Yes" : "No");
  }
}

void main() {
  var provider = new TestIOProvider();
  new Program(provider).main();
}
